daemon/emer-persistent-cache.c

```diff
@@ -1664,38 +868,210 @@ emer_persistent_cache_initable_init (GInitableIface *iface)
+ * its value.
+ *
+ * Returns TRUE on success. Returns FALSE on failure and sets the error unless
+ * it's NULL. Does not alter the offset out parameter on failure.
+ */
+gboolean
+emer_persistent_cache_get_boot_time_offset (EmerPersistentCache *self,
+                                            gint64              *offset,
+                                            gboolean             always_update_timestamps,
+                                            GError             **error)
+{
+  EmerPersistentCachePrivate *priv =
+    emer_persistent_cache_get_instance_private (self);
+
+  /*
+   * When always_update_timestamps is FALSE, the timestamps won't be written
+   * unless the boot offset in the metadata file is being overwritten.
+   */
+  gboolean update_succeeded =
+    update_boot_offset (self, always_update_timestamps, error);
+  if (!update_succeeded)
+    return FALSE;
+
+  if (offset != NULL)
+    *offset = priv->boot_offset;
+
+  return TRUE;
+}
+
+
+/* Persistently stores the given variants. Sets num_variants_stored to the
+ * number of variants that were actually stored. Returns TRUE on success even
+ * if all of the given variants don't fit in the space allocated to the
+ * persistent cache. Returns FALSE only on error.
+ */
+gboolean
+emer_persistent_cache_store (EmerPersistentCache *self,
+                             GVariant           **variants,
+                             gsize                num_variants,
+                             gsize               *num_variants_stored,
+                             GError             **error)
+{
+  EmerPersistentCachePrivate *priv =
+    emer_persistent_cache_get_instance_private (self);
+
+  gsize curr_variants_stored = 0;
+  for (; curr_variants_stored < num_variants; curr_variants_stored++)
+    {
+      GVariant *curr_variant = variants[curr_variants_stored];
+      GVariant *copy = deep_copy_variant (curr_variant);
+      g_variant_ref_sink (copy);
+      GVariant *little_endian_variant = swap_bytes_if_big_endian (copy);
+      g_variant_unref (copy);
+      GVariant *normalized_variant =
+        g_variant_get_normal_form (little_endian_variant);
+      g_variant_unref (little_endian_variant);
+
+      const gchar *type_string = g_variant_get_type_string (normalized_variant);
+      gsize type_length = strlen (type_string) + 1;
+      gconstpointer variant_data = g_variant_get_data (normalized_variant);
+      gsize variant_size = g_variant_get_size (normalized_variant);
+
+      gsize elem_size = type_length + variant_size;
+      guint8 elem[elem_size];
+      memcpy (elem, type_string, type_length);
+      memcpy (elem + type_length, variant_data, variant_size);
+      g_variant_unref (normalized_variant);
+      if (!emer_circular_file_append (priv->variant_file, elem, elem_size))
+        break;
+    }
+
+  if (!emer_circular_file_save (priv->variant_file, error))
+    return FALSE;
+
+  for (gsize i = 0; i < curr_variants_stored; i++)
```

Comment:
The daemon ref sinks its variants, so I don't see where these would be doubly unreffed. The caller should unref the variants if they took a reference on them but not if they didn't. I believe this is the standard behavior of most APIs in glib.